BUG: Overflow in tan and tanh for large complex values

charris opened the pull request:
Rebase of #3010.

`np.tanh(1000+0j)` gives `nan+nan*j` instead of `1.0+0j`.
`np.tan(0+1000j)` gives  `nan+nan*j` instead of `1j`.

I've imported the implementation for ctanh from FreeBSD's math library
which handles large arguments correctly and fixes this bug and the
equivalent bug in np.tan.

The problem here is that importing the function into npy_math_complex.c
and adding a config check causes us to use the implementation from glibc
on Linux, which also has this bug. Although it seems to have been fixed
in glibc last April
(http://sourceware.org/bugzilla/show_bug.cgi?id=11521).

I see several things that could be done here, the easiest is probably to
use our version of ctanh unconditionally. Although there are a multitude
of ways to go about that, for instance should I remove the
implementation from npy_math_complex.c and just place it directly in
umath/funcs.inc, where the buggy version was? Or should I just remove
the config check and add a note about it somewhere?

Closes #2321.

ewmoore committed:
BUG: Overflow in tan and tanh for large complex values

np.tanh(1000+0j) gives nan+nan*j instead of 1.0+0j.
np.tan(0+1000j) gives  nan+nan*j instead of 1j.

I've imported the implementation for ctanh from FreeBSD's math library
which handles large arguments correctly and fixes this bug and the
equivalent bug in np.tan.

The problem here is that importing the function into npy_math_complex.c
and adding a config check causes us to use the implementation from glibc
on Linux, which also has this bug. Although it seems to have been fixed
in glibc last April
(http://sourceware.org/bugzilla/show_bug.cgi?id=11521).

I see several things that could be done here, the easiest is probably to
use our version of ctanh unconditionally. Although there are a multitude
of ways to go about that, for instance should I remove the
implementation from npy_math_complex.c and just place it directly in
umath/funcs.inc, where the buggy version was? Or should I just remove
the config check and add a note about it somewhere?

Closes #2321.

charris commented:
Squashed commits of #3010 and rebased on master. @juliantaylor Might want to rebase this for backport pending review.

ewmoore commented:
One aspect of this that was never really resolved were the build system changes. This compiles and executes some tests to decide whether to use our internal function or the libc version. Which I guess is fine but never really seemed ideal.

charris commented on numpy/core/code_generators/generate_umath.py:
```diff
@@ -33,7 +33,7 @@ class TypeDescription(object):
     type : str
         Character representing the nominal type.
-    func_data : str or None or FullTypeDescr, optional
+    func_data : str or None or FullTypeDescr or FuncNameSuffix, optional
```
Hmm, might want to remove this change.

charris replied:
`FuncNameSuffix` is used in master, so keep this.

charris commented:
Needs fixups.

1) C style fixups
2) inline -> NPY_INLINE
3) probably more...

charris commented on numpy/core/src/umath/umathmodule.c:
```diff
@@ -370,6 +371,11 @@ PyMODINIT_FUNC initumath(void)
     Py_DECREF(s);
 
+#if defined(NPY_PY3K)
+    s = PyDict_GetItemString(d, "true_divide");
```
Do we want to do this?

charris replied:
Things were working before this, why make this change?

ewmoore replied:
These lines have just been moved. This whole #if was added in cfc5c3bc0c.

charris committed:
STY: C style cleanup.

Omitted numpy/core/src/npymath/test_c99complex.c, which is a jungle
of macro magic.

charris commented on numpy/core/setup.py:
```diff
@@ -209,6 +209,32 @@ def check_prec(prec):
             else:
                 priv.extend([(fname2def(f), 1) for f in flist])
 
+            flist = [f + prec for f in C99_COMPLEX_FUNCS_CHECKED]
```
Makes me wonder if these tests shouldn't be done in python.

ewmoore replied:
There was some discussion of using ctypes to call these functions from python.  I didn't go down that route since we then need to find libc ourselves, but doing it this way allowed the tool chain to find it for us.  I suppose we could make a tiny module that simply exposes them to python and then use that for the tests.

ewmoore commented on numpy/core/include/numpy/npy_math.h:
```diff
@@ -258,8 +261,11 @@ float npy_copysignf(float x, float y);
 float npy_nextafterf(float x, float y);
 float npy_spacingf(float x);
 
+float npy_ldexpf(float x, int exp);
```
npy_ldexp and npy_frexp in all three flavors already exist in master.  These came in as part of #4852, which pulled the ldexp and frexp changes from this.

ewmoore commented on numpy/core/src/npymath/npy_math_complex.c.src:
```diff
@@ -31,9 +34,11 @@
-/*==========================================================
- * Custom implementation of missing complex C99 functions
- *=========================================================*/
+#define raise_inexact() do { volatile npy_float junk = 1 + tiny; } while(0)
```
I was always temped to remove this entirely.  The inexact exception is not exposed to python AFAICT.

charris commented:
@ewmoore I have a simplified version of this PR. Given the problems with csinh in gcc < 4.8, I'm inclined to simply restrict detection of system libraries to the current set and use your new routines for the rest. Other sources of these library routines that we could use in the future come with the languages D (Boost license) and GO (three clause BSD), both of which would need translation to C. The D library seems to have better support of quad precision (SUN).  It is difficult to understand how gcc could have gotten things wrong the way they did.

We could also simply blacklist some compilers, but the maintenance burden of that is likely too heavy.

charris commented:
Closing this. @ewmoore #5518 is a simplified version that uses a black list instead of tests. While I agree that tests are probably the most reliable way to check unknown architectures, the simplicity of a black list has much to recommend it, especially if the problem is mostly older libraries.
